### DISCUSSION TITLE
How to convert (ugly, dumb) raw SQL in SQLAlchemy ORM(or Core) in right way

### DISCUSSION THREAD

**Discussion by capymind** [sqlalchemy/sqlalchemy#9906] [Usage Questions] [answered]
It's not directly related to specific topic in SQLAlchemy. 
(Sorry for not-detailed(specific) but just naive and general question..)
However, I want to listen from people who has many years of doing SQLAlchemy.

When trying to solve some complex problem, the way raw SQL is written is naturally come up with my head.
Definitely the first draft is not good at all (even terrible!) but works in the way I intended, meaning get right rows.

For example, this dumb query:

```sql
SELECT *
FROM
(
    SELECT    A.*
            , B.id  AS in_database
    FROM
    (
        SELECT *
        FROM opendart_metadata
        WHERE 1=1
            AND id IN (
                            SELECT id
                            FROM
                            (
                                SELECT id,
                                       ROW_NUMBER() OVER(
                                            PARTITION BY report_type, report_period, report_year 
                                            ORDER BY report_zipname DESC
                                        ) AS seq
                                FROM opendart_metadata
                                WHERE 1=1
                                    AND report_year IN (2021, 2022)
                            ) AS A
                            WHERE seq = 1 
                        )
    ) AS A
    LEFT JOIN financial_statement_metadata AS B
    ON 1=1
        AND A.report_zipname == B.report_zipname
) AS A
WHERE in_database IS NULL
```

Whenever I try to convert this kind of messy query into SQLAlchemy way using `select()`, `join`, `row_number().over()`, `subquery()`, `aliased()` etc.. I am stuck with the fact that I cant write any line of code.. As a workaround just using `text()` with `bindparms` saves me.

Is there any general approach(recommendation) to structure sqlalchemy's code to convert messay(ugly) raw sql into python code?
I think the cliche like practice, practice and practice is correct tough answer..  Anyway thanks for reading my silly discussion..

SQLAlchemy itself is awesome and I really appreciate maintainors for managing this good projects even better!

**Comment by CaselIT** [comment, marked as answer]
> Whenever I try to convert this kind of messy query into SQLAlchemy way using `select()`, `join`, `row_number().over()`, `subquery()`, `aliased()` etc.. I am stuck with the fact that I cant write any line of code.. 
I'm not sure what you mean by this

You generally start from the inner subquery and build your way to the external one.
You should better define the tables, but this seems to be the rough equivalent
```py
import sqlalchemy as sa

om = sa.table(
    "opendart_metadata",
    sa.column("id"),
    sa.column("report_type"),
    sa.column("report_period"),
    sa.column("report_year"),
    sa.column("report_zipname"),
)
fsm = sa.table(
    "financial_statement_metadata",
    sa.column("id"),
    sa.column("report_zipname"),
)

s1 = (
    sa.select(
        om.c.id,
        sa.func.row_number()
        .over(
            [om.c.report_type, om.c.report_period, om.c.report_year],
            om.c.report_zipname,
        )
        .label("seq"),
    )
    .where(om.c.report_year.in_((2021, 2022)))
    .subquery("A")
)

s2 = (
    sa.select(om)
    .where(om.c.id.in_(sa.select(s1.c.id).where(s1.c.seq == 1)))
    .subquery("A")
)

s3 = (
    sa.select(s2, fsm.c.id.label("in_database"))
    .select_from(s2)
    .join(fsm.alias("B"), s2.c.report_zipname == fsm.c.report_zipname)
    .subquery("A")
)

stmt = sa.select("*").select_from(s3).where(s3.c.in_database.is_(sa.null()))

print(stmt)
```

**Reply by capymind** [reply]
Thank you a lot. The code itself and your saying are helpful a lot for me :)
> You generally start from the inner subquery and build your way to the external one.
